File: packages/python/src/transmute/body2d.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


class Solid(Protocol):
    @property
    def x(self) -> float: ...

    @property
    def y(self) -> float: ...

    @property
    def width(self) -> float: ...

    @property
    def height(self) -> float: ...


@dataclass
class Body2D:
    x: float
    y: float
    width: float
    height: float
    velocity_x: float = 0.0
    velocity_y: float = 0.0
    on_ground: bool = False
    gravity: float = 0.5
    jump_strength: float = -10.0

    def set_velocity_x(self, vx: float) -> None:
        self.velocity_x = vx

    def jump(self) -> None:
        if self.on_ground:
            self.velocity_y = self.jump_strength
            self.on_ground = False

    def step(self, solids: list) -> None:
        self.velocity_y += self.gravity
        self.x += self.velocity_x
        self._resolve_x(solids)
        self.y += self.velocity_y
        self.on_ground = False
        self._resolve_y(solids)

    def _resolve_x(self, solids: list) -> None:
        for s in solids:
            if self._aabb(s):
                if self.velocity_x > 0:
                    self.x = s.x - self.width
                elif self.velocity_x < 0:
                    self.x = s.x + s.width
                self.velocity_x = 0

    def _resolve_y(self, solids: list) -> None:
        for s in solids:
            if self._aabb(s):
                if self.velocity_y > 0:
                    self.y = s.y - self.height
                    self.velocity_y = 0
                    self.on_ground = True
                elif self.velocity_y < 0:
                    self.y = s.y + s.height
                    self.velocity_y = 0

    def _aabb(self, s) -> bool:
        return (
            self.x < s.x + s.width
            and self.x + self.width > s.x
            and self.y < s.y + s.height
            and self.y + self.height > s.y
        )

    @property
    def is_on_ground(self) -> bool:
        return self.on_ground

File: packages/java/transmute-core/TransmuteCore/src/TransmuteCore/physics/Body2D.java
package TransmuteCore.physics;

import TransmuteCore.math.Collision;

import java.util.List;

/**
 * Deep 2D rigid body for platformers: gravity, jump, AABB resolve against solids.
 */
public final class Body2D
{
    public interface Solid
    {
        float getX();

        float getY();

        float getWidth();

        float getHeight();
    }

    private float x;
    private float y;
    private float velocityX;
    private float velocityY;
    private final float width;
    private final float height;
    private boolean onGround;
    private float gravity = 0.5f;
    private float jumpStrength = -10f;

    public Body2D(float x, float y, float width, float height)
    {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    public Body2D gravity(float gravity)
    {
        this.gravity = gravity;
        return this;
    }

    public Body2D jumpStrength(float jumpStrength)
    {
        this.jumpStrength = jumpStrength;
        return this;
    }

    public void setVelocityX(float velocityX)
    {
        this.velocityX = velocityX;
    }

    public void setPosition(float x, float y)
    {
        this.x = x;
        this.y = y;
    }

    public void jump()
    {
        if (onGround)
        {
            velocityY = jumpStrength;
            onGround = false;
        }
    }

    /**
     * Integrates velocity and resolves against {@code solids}.
     */
    public void step(List<? extends Solid> solids)
    {
        if (!onGround)
        {
            velocityY += gravity;
        }
        x += velocityX;
        y += velocityY;
        resolve(solids);
    }

    private void resolve(List<? extends Solid> solids)
    {
        onGround = false;
        for (Solid solid : solids)
        {
            Collision.Resolution hit = Collision.resolveAabb(
                x, y, width, height,
                solid.getX(), solid.getY(), solid.getWidth(), solid.getHeight());
            if (hit.collided)
            {
                x = hit.x;
                y = hit.y;
                if (hit.landed())
                {
                    velocityY = 0;
                    onGround = true;
                }
                else if (hit.hitBottom)
                {
                    velocityY = 0;
                }
                else if (hit.hitHorizontal)
                {
                    velocityX = 0;
                }
            }
        }
        if (!onGround)
        {
            for (Solid solid : solids)
            {
                if (Collision.aabb(
                    x, y + 1, width, height,
                    solid.getX(), solid.getY(), solid.getWidth(), solid.getHeight()))
                {
                    y = solid.getY() - height;
                    velocityY = 0;
                    onGround = true;
                    break;
                }
            }
        }
    }

    public float getX()
    {
        return x;
    }

    public float getY()
    {
        return y;
    }

    public float getWidth()
    {
        return width;
    }

    public float getHeight()
    {
        return height;
    }

    public float getVelocityY()
    {
        return velocityY;
    }

    public boolean isOnGround()
    {
        return onGround;
    }
}
